src/core/ext/transport/chttp2/server/chttp2_server.cc

```diff
@@ -289,6 +289,55 @@ static void server_destroy_listener(grpc_server* server, void* arg,
   grpc_tcp_server_unref(tcp_server);
 }
 
+static grpc_error* chttp2_server_add_acceptor(grpc_server* server,
+                                              const char* name,
+                                              grpc_channel_args* args) {
+  grpc_tcp_server* tcp_server = nullptr;
+  grpc_error* err = GRPC_ERROR_NONE;
+  server_state* state = nullptr;
+  const grpc_arg* arg = nullptr;
+  grpc_core::TcpServerFdHandler** arg_val = nullptr;
+
+  state = static_cast<server_state*>(gpr_zalloc(sizeof(*state)));
+  GRPC_CLOSURE_INIT(&state->tcp_server_shutdown_complete,
+                    tcp_server_shutdown_complete, state,
+                    grpc_schedule_on_exec_ctx);
+  err = grpc_tcp_server_create(&state->tcp_server_shutdown_complete, args,
+                               &tcp_server);
+  if (err != GRPC_ERROR_NONE) {
+    goto error;
+  }
+
+  state->server = server;
+  state->tcp_server = tcp_server;
+  state->args = args;
+  state->shutdown = true;
+  gpr_mu_init(&state->mu);
+
+  // TODO(yangg) channelz
+
+  arg = grpc_channel_args_find(args, name);
```

Comment:
we are using the address name as the channel arg?

Reply:
Yes. I am using the "external: x" as a key to reference a pointer space here. The issue is that without this hack, we cannot return anything from this layer to the upper layer. So this is the way to return a handle to the actual connection handler from the tcp layer to the external acceptor to enable the connection hand-off.